man/man1/tabtos.1
.TH TABTOS 1NEMO "18 January 2012"
.SH NAME
tabtos \- convert structured ASCII tables into snapshot format
.SH SYNOPSIS
\fBtabtos in=\fPascii_file \fBout=\fPsnap_file [parameter=value] .\|.\|.
.SH DESCRIPTION
\fItabtos\fP converts ASCII tables into
\fIsnapshot(5NEMO)\fP format. These tables can be structured as well
as unstructured tables. There is a control parameter
(\fBoptions=scan\fP) with which one 
may be able to figure out the structure of a
file. For binary files, there are several examples,
as well the more generic \fIbinsnap(1NEMO)\fP.
.PP
Structured tables consist of
a small (though optional) header, 
that normally contains parameters like the number of bodies,
the time of the snapshot etc., and one or more blocks (sections)
of homogeneous tables, their length equal the number of bodies
in the snapshot and their width as specified by the user
from the variables expected to be present in those blocks. Examples
will be shown below.
.PP
Unstructured files (i.e. without a header) can also be
processed if the user specifies how many bodies are present per
snapshot.  If even that is
not done, the number of lines in the file (corrected for the
number of \fBtimes\fP and \fBblock\fPs 
specified) is used to derive the number of
bodies.  Although at least one block must be present, if no
times are specified, the number of snapshots is then assumed to
be one. See examples below. 
.PP
By default tabtos does not handle comment lines (lines starting with
#) gracefully at all,
see \fBoptions=comment\fP or \fBoptions=history\fP below how to deal with his.
If you use these options, automatically determining \fBnbody\fP will
not work, and either \fBheader=\fP or \fBnbody=\fP needs to be used.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBin=\fP\fIascii_file\fP
Input file name (ascii table/structure)
No default.
.TP
\fBout=\fP\fIsnap_file\fP
Output file name (snapshot).
No default.
.TP
\fBheader=\fP\fIvar1,var2,...\fP
A list of variables that are present in the header, in
the order given. They do not need to be present on the same line.

Recognized are: \fBnbody, ndim, time, skip (-)\fP. If they do not
appear in the header, their value should be set via the
equivalent keyword (see below). The special value \fBskip\fP (or \fB-\fP) 
is allowed to skip header items that have no current meaning in
parsing the data into snapshot format (e.g. SPH files have 
\fBngas\fP, but this is currently not processed).
Default: none supplied.
.TP 20
\fBblock1=\fP\fIvar1,var2,...\fP
A list of variables that are present in this block 
of \fBnbody\fP lines, in the order give.

Recognized are: \fBmass, pos, vel, phase, acc, phi, dens,
aux\fP for real-valued and \fBkey\fP for integer valued
columns (see also \fIsnapshot(5NEMO)\fP). In addition, single
elements from the \fBndim\fP vector are recognized as
\fBx, y, z, vx, vy, vz, ax, ay, az\fP. \fBm\fP is also
recognized as \fBmass\fP, to be symmetric with the
standard \fIbodytrans(5NEMO)\fP names.
As before, the name \fBskip\fP (or \fB-\fP) can be used to skip a column.
If the variable \fBskip\fP appears are the only item in a block,
all \fBnbody\fP rows in that block are skipped, independant of 
the number of columns (NOTE: this cannot be done if
\fBoptions=wrap\fP is used). 

Structured ASCII tables can contain up to 9 separate blocks,
currently all of them must have the same length, and be equal to
the number of bodies, \fBnbody\fP.
No default.
.TP
\fBblock2=\fP\fIvar1,var2,...\fP
See above, for 2nd block, if present.  
Default: not used.
.TP
\fBblock3=\fP\fIvar1,var2,...\fP
See above, for 3rd block, if present.  
Default: not used.
.TP
\fBblock4=\fP\fIvar1,var2,...\fP
See above, for 4th block, if present.
Default: not used.
.TP
\fBblock5=\fP\fIvar1,var2,...\fP
See above, for 5th block, if present.
Default: not used.
.TP
\fBblock6=\fP\fIvar1,var2,...\fP
See above, for 6th block, if present.
Default: not used.
.TP
\fBblock7=\fP\fIvar1,var2,...\fP
See above, for 7th block, if present.
Default: not used.
.TP
\fBblock8=\fP\fIvar1,var2,...\fP
See above, for 8th block, if present.
Default: not used.
.TP
\fBblock9=\fP\fIvar1,var2,...\fP
See above, for 9th block, if present.
Default: not used.
.TP
\fBnbody=\fP\fInbody\fP
Number of bodies, if the header doesn't supply it. If none supplied, neither
through the header, nor using this keyword, it is derived from
the number of lines in the file, corrected (divided) by
the number of specified blocks (\fBblock#=\fP)
and the numbers of specified times. 
(\fBtimes=\fP).
Default: not used.
.TP 20
\fBndim=\fP\fINDIM\fP
Dimension of \fBpos,vel,acc\fP arrays, if the header doesn't supply it.
The program can be compiled for \fINDIM\fP=2 or 3, but they
must match. If not specified, the program will assume it's internal;
default, \fINDIM\fP.
Default: not used.
.TP 20
\fBtimes=\fP\fIt1,t2,...\fP
Time(s) of snapshot(s), if the header doesn't supply it.
This can be a list of times, if multiple snapshot are being read.
Implied do loops can be given, for example, 
\fB0:100:2.5\fP is a legal way to specify an array of 41 times
(see \fInemoinp(1NEMO)\fP)
Specifying times through this keyword not only overrides any
header values, but also limits the maximum number of snapshot
that are read (thus avoiding the dreaded fatal EOF error)
One can also use \fBoptions=time\fP,
in which case time is auto-incremented (by 1) for each snapshot.
Default: not used.
.TP
\fBoptions=\fP
A list of options, needs exact match.
\fBscan\fP the input file is merely scanned for
regularity. It then prints the line numbers at which the
number of columns changed. It's a bit more informational than
\fIawk '{print NF}' | uniq\fP.
\fBcomment\fP allows you to have comment lines in the file (lines
that start with a # symbol). By default no comment lines are allowed.
\fBwrap\fP allows, for each block, 
the expected numbers on a line to be wrapped over more than
one line. The side-effect of this is that the next record in that
same block starts immediately following the wrapped number. 
\fBspill\fP discards all numbers read from the last line to fill a record
within a block, and hence negates the side-effect of the \fPwrap\fP option.
\fBtime\fP auto-increments time (by 1) as new snapshots are found.
\fBhistory\fP uses comment lines and adds them to the data history
(see \fIhistory(3NEMO)\fP).
.TP
\fBheadline=\fP
Random mumblage for humans, used to identify the snapshot.
Default: not used.
.SH EXAMPLES
To read in a single snapshot table with masses, (3D) positions and 
velocities in columns 1-7, with the number of bodies equals the number
of rows in the table, you can use:
.nf

    % tabtos in=table out=snap block1=mass,pos,vel

.fi
To read in "205" formatted N-body snapshots 
(see also \fIatos(1NEMO)\fP), with three seperate blocks of
masses, positions and velocities, you can use:
.nf

    % tabtos in=table out=snap header=nbody,ndim,time \\
            block1=mass block2=pos block3=vel

.fi
To read in "205" pure-SPH data (i.e. NGAS=NOBJ, see also 
\fIatosph(1NEMO)\fP), in which the
density is going to be
stored in the \fBmass\fP slot, entropy/temperature
in \fBaux\fP, potentials in \fBphi\fP, and masses and smoothing
lengths are skipped, use:
.nf

    % tabtos in=table out=snap header=nbody,skip,ndim,time  \\
            block1=skip block2=pos block3=vel block4=mass block5=aux \\
            block6=skip block7=phi

.fi
Note: files with Ngas < Nobj cannot be processed (yet), since the
blocks have different length. See \fIatosph(1NEMO)\fP.
.SH 205 FORMAT
For a full explanation, see \fIatos(1NEMO)\fP, but below we list
the two different popular "205" formats around (atos, and atosph)

.nf
.ta +1i +2i
\fIformat	header\fP          	blocks\fP

atos	nbody,ndim,time  	mass pos vel phi
atosph	nbody,ngas,ndim,time	mass pos vel rho temp hsph phi acc
.fi
.SH CAVEATS
\fItabtos\fP is very tolerant (sometimes too!)
with respect to slight
under- or over-specifications. It tries to write data, but only
if it really gets stuck, the program is aborted. 
\fIwarning(3NEMO)\fP messages should not be discarded, check
them to see if they make sense. Otherwise what you think is
correct data, may not have been parsed correctly. That's life
in the fast ASCII lane. 
.PP
If no \fBtimes\fP specified, and the \fBheader=\fP has no 
associated time, all snapshots will 
have the same time (0.0). A warning will be issued.
.PP
If both \fBtimes\fP specified, and the \fBheader=\fP has an
associated time, the header value will be overriden with
the user specified values. Note that the number of specified
times now determines the maximum number of snapshots that
will be read, i.e. EOF may not be reached.
.PP
Although snapshots with varying amounts of particles can be
read, there are many programs in NEMO which have difficulties
if the first snapshot is not the largest one in that file.
.SH BUGS
This is a complicated program to understand, but can often read complicated
ascii files.
.PP
This program cannot parse fixed column files if they are not separated by whitespace.
.PP
Files with comment lines cannot be parsed. Use \fItabcomment(1NEMO)\fP to
delete them:
.nf
    tabcomment tab.in - delete=t | tabtos - snap.out ....
.fi
though the drawback of using a pipe in this example is that typically 
the keyword \fBnbody=\fP needs to be supplied.
.SH SEE ALSO
atos(1NEMO), atosph(1NEMO), binsnap(1NEMO), snapprint(1NEMO), snapshot(5NEMO), unfio(1NEMO)
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
27-Aug-93	V1.0 Created, I finally broke down	Peter Teuben
30-aug-93	V1.1 added scan=                   	PJT
25-oct-94	V1.2 options=scan|comment	PJT
2-nov-94	V1.2d added wrap/spill, fixed times= bug	PJT
19-aug-00	V1.3d fixed various TAB related problems	PJT
24-jan-02	V1.3f block10 now	PJT
29-jul-05	V1.4 added options=time to auto-inc time	PJT
14-nov-06	V1.5 added history option			PJT
18-jan-12	V1.5a  added dens(ity) option	JCL
.fi


src/nbody/io/tabtos.c
/*
 *  TABTOS - generalized tables to snapshot converter
 *
 *           atos(ph) was too cumbersome, as shown by the following 
 *           sequence to read in pure SPH data:
 *
 *           (in, out, var and color are useful shell variables)
 *
 *               atosph $in - | tsf - allline=t |\
 *               sed s/$var/Aux/g  | rsf - - |\
 *               snapcenter - - tab=f |\
 *               snapxyz - $out "color=$color" 
 *
 *           moreover, tabtos is 2.0 times faster than atosph,
 *		probably due to fscanf()  vs. fgets()+nemoinp()
 *
 *      26-aug-93   V1.0    Created - while in Tokyo           PJT
 *      30-aug-93   V1.1    Added options=scan                 PJT
 *	10-dec-93   V1.2    renamed time= to times=	       pjt
 *	22-feb-94   V1.2a   ansi headers
 *	 2-sep-94   V1.2b   fixed bug in get_double(), didn't return #
 *	26-oct-94   V1.2c   added options=comment 	pjt
 *       2-nov-94   V1.2d   added options=wrap,spill for Sellwood	pjt
 *			    (@Rutgers)
 *	13-dec-94   V1.2f   patch nbody if not enough nbody could be read pjt
 *			    (needs error=1)
 *	15-mar-95   V1.3    allow x,y,z   vx,vy,vz for pos,vel		pjt
 *	 3-aug-95   V1.3a   allow TABs in files				pjt
 *	25-mar-97	b   SINGLEPREC fix				pjt
 *       7-jul-98       c   fixed bug skipping blocks                   pjt
 *      19-aug-00       d   fixed TABs correctly                        pjt
 *	23-sep-01       e   ->nemo_file_lines
 *      24-jan-02       f   to block10, time to use indexed parameter   pjt
 *       3-feb-02       g   fixed bug if time in header, nbody not      pjt 
 *      11-may-02       h   longer lines to deal with DCR's 'ss2bt' files   PJT
 *      28-may-03       i   allow 'm' also for 'mass', also allow - for skip  pjt
 *      21-sep-03       j   fixed bug in calling nemo_file_lines              pjt
 *      29-jul-05   V1.4    added auto-incrementing time option (Amsterdam, cafe amaricain) pjt
 *      14-nov-06   V1.5    add first # comments to the NEMO output history.
 *      18-Jan-12   V1.5a   add 'dens' array                           jcl
 */

#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>

#include <snapshot/snapshot.h>
#include <snapshot/body.h>
#include <snapshot/put_snap.c>

string defv[] = {
    "in=???\n      Input file name - ascii (structured) table",
    "out=\n        Output file name - snapshot",
    "header=\n     Header scalars (nbody, ndim, time, skip, -)",
    "block1=\n     What's in first block of nbody lines - mass",
    "block2=\n     - pos(ndim)",
    "block3=\n     - vel(ndim)",
    "block4=\n     - phase(2*ndim)",
    "block5=\n     - acc(ndim)",
    "block6=\n     - phi",
    "block7=\n     - aux",
    "block8=\n     - key",
    "block9=\n     - skip",
    "block10=\n    - skip",
    "nbody=\n      Number of bodies, if needed",
    "ndim=\n	   Dimension of pos,vel,acc arrays, if needed",
    "times=\n      Time(s) of snapshot, if needed, or override",
    "options=\n    Other processing options (scan|comment|wrap|spill|time)",
    "nskip=0\n     Number of lines skipped before each (header+block1+...)",
    "headline=\n   Random mumblage for humans",
    "VERSION=1.5b\n 18-Jan-2012 PJT",
    NULL,
};

string usage = "convert ASCII tabular information to snapshot format";

string cvsid = "$Id$";


#define MAXLINE  1024
#define MAXTIMES 1024
#define MAXVALS    32
#define MAXHIST   128

local stream instr, outstr;

local int nbody=0, nbody_max=0, bits=0;
local real tsnap=0.0;
local Body *btab = NULL;

local string *header;
local int nheader=0;

#define MAXBLOCK 10
local string block[MAXBLOCK];
local int nblocks=0;

local real times[MAXTIMES];
local int ntimes=0;

local int nskip=0;
local int linecnt=0;
local int snapcnt=0;
local bool scan, wrap, spill, auto_time, Qhis, Qcom;


local int nhist = 0;
local string shist[MAXHIST];

/* extern's */
extern string *burststring(string, string);
extern bool scanopt(string, string);

/* local's */
local void check_options(void);
local bool get_header(void);
local bool get_block(int, string );
local int get_double(int, double *);
local int get_nbody(void);
local void do_scan(stream);     
local void tab2space(string);

void nemo_main(void)
{
    int i;
    char name[20];
    bool ok;

    check_options();

    header = burststring(getparam("header"),", ");
    nheader = xstrlen(header,sizeof(string)) - 1;

    for (i=0; i<MAXBLOCK; i++) {
        sprintf(name,"block%d",i+1);
        block[i] = getparam(name);
        if (hasvalue(name))
	    nblocks++;
        else
            break;
    }
    /* additional error check if not in scan mode */
    if (!scan) {
        if (!hasvalue("out")) error("Need output filename; out=");
        if (nblocks==0) error("Need at least one block1=; maximum is block9=");
    }

    ntimes = nemoinpr(getparam("times"),times,MAXTIMES);
    if (ntimes < 0) error("Parsing times=, perhaps too many? (<%d)",MAXTIMES);

    instr = stropen(getparam("in"), "r");
    if (scan) {
        do_scan(instr);
        strclose(instr);
        stop(0);
    }
    
    outstr = stropen(getparam("out"), "w");
    if (hasvalue("headline")) set_headline(getparam("headline"));

    while (get_header()) {	/* loop reading snapshots */
        snapcnt++;
        if (nbody==0) {         /* if unknown, try and estimate  */
            nbody = get_nbody();
            dprintf(0,"Assuming nbody=%d\n",nbody);
        }
        if (ntimes > 0) {		/* override or set time */
            if (snapcnt <= ntimes)
                tsnap = times[snapcnt-1];
            else
                warning("Too few snapshot times= specified, good luck");
        }
	if (auto_time)
	  tsnap += 1.0;
        dprintf(0, "[reading %d bodies at time %f]\n", nbody, tsnap);
        if (btab==NULL) {
            btab = (Body *) allocate(nbody*sizeof(Body));
            nbody_max = nbody;
	}
        if (nbody > nbody_max) {
            warning("Reallocating snapshot from %d to %d",nbody,nbody_max);
            btab = (Body *) reallocate(btab,nbody*sizeof(Body));
            nbody_max = nbody;
        }
        if (nbody <= 0) {
            warning("Bad nbody, possibly wrong number of block's specified");
            break;
        }

        bits = TimeBit;			/* reset the bits */
        for (i=0; i<nblocks; i++) {	/* process for each block */
            ok = get_block(i+1,block[i]);
            if (!ok) {
		break;
	    }
        }
        if (ok) {
	  if (nhist > 0) {
	    for (i=0; i<nhist; i++)
	      app_history(shist[i]);
	    nhist = 0;
	  }
	  put_history(outstr);
	  put_snap(outstr, &btab, &nbody, &tsnap, &bits);	      /* output */
        } else
	  break;
        if (ntimes > 0 && snapcnt == ntimes) break;
    }
    strclose(outstr);
} /* nemo_main */

void check_options(void)
{
    string options;
    bool scanopt();

    if (!hasvalue("options")) return;

    options=getparam("options");
    scan = scanopt(options,"scan");
    Qcom = scanopt(options,"comment");
    wrap = scanopt(options,"wrap");
    spill = scanopt(options,"spill");
    auto_time = scanopt(options,"time");
    Qhis = scanopt(options,"history");
    if (Qhis) Qcom = TRUE;

    nskip = getiparam("nskip");
}

local bool get_header(void)
{
    int i, ndim=0;
    double dval, *dvals;
    char line[MAXLINE];

    if (feof(instr)) return FALSE;

    /* @TODO:  nskip appears not to work??? */
    for (i=0; i<nskip; i++) {
        if (fgets(line,MAXLINE,instr) == NULL) {
            if (i==0) return FALSE;
            error("in_header(%d): unexpected EOF at line %d",i+1,linecnt);
        }
    }

    dvals = (double *) allocate(nheader * sizeof(double));

    if (nheader > 0) {
        if (get_double(nheader,dvals) != nheader) return FALSE;
        for (i=0; i<nheader; i++) {
            if (streq(header[i],"nbody"))
                nbody = dvals[i];
            else if (streq(header[i],"time"))
                tsnap = dvals[i];
            else if (streq(header[i],"ndim"))
                ndim = dvals[i];
            else if (streq(header[i],"skip"))
                dval = dvals[i];
            else if (streq(header[i],"-"))
                dval = dvals[i];
            else
                warning("Skipping unknown header name %s",header[i]);
        }
    } else
	if (feof(instr)) return FALSE;

    if (nbody==0) {
        if (hasvalue("nbody")) nbody=getiparam("nbody");
        ndim = hasvalue("ndim") ? getiparam("ndim") : NDIM;
    }

    if (ndim>0 && ndim != NDIM)
	error("got ndim = %d, not %d", ndim, NDIM);
    return TRUE;
}

local bool get_block(int id,string options)
{
    int i, j, n, need, ngot, nvals;
    char line[MAXLINE];
    double dvals[MAXVALS];
    string *o;
    Body *bp;
    bool skip = streq(options,"skip");   /* if to skip the whole block */
    int colmass, colpos, colvel, colphase, colphi, colacc, colaux, coldens, colkey;
    int colx, coly, colz, colvx, colvy, colvz, colax, colay, colaz;

    dprintf(1,"Block%d: %s",id,options);
    
    if (options==NULL || *options==0) return FALSE;
    o = burststring(options,", \t");
    n = xstrlen(o,sizeof(string)) - 1;
    colmass=colpos=colvel=colphase=colphi=colacc=colaux=colkey=coldens=0;
    colx=coly=colz=colvx=colvy=colvz=colax=colay=colaz=0;

    need = 0;
    for (i=0; i<n; i++) {
        if (streq(o[i], "skip") || streq(o[i], "-")) {
            need += 1;
        } else if (streq(o[i], "mass")) { 
            colmass = need+1;   need += 1;      bits |= MassBit;
        } else if (streq(o[i], "m")) { 
            colmass = need+1;   need += 1;      bits |= MassBit;
        } else if (streq(o[i], "phase")){ 
            colphase= need+1;   need += 2*NDIM; bits |= PhaseSpaceBit;
	} else if (streq(o[i], "pos"))  { 
            colpos  = need+1;   need += NDIM;   bits |= PhaseSpaceBit;
	} else if (streq(o[i], "x"))  { 
            colx    = need+1;   need += 1;      bits |= PhaseSpaceBit;
	} else if (streq(o[i], "y"))  { 
            coly    = need+1;   need += 1;      bits |= PhaseSpaceBit;
	} else if (streq(o[i], "z"))  { 
            colz    = need+1;   need += 1;      bits |= PhaseSpaceBit;
	} else if (streq(o[i], "vel"))  { 
            colvel  = need+1;   need += NDIM;   bits |= PhaseSpaceBit;
	} else if (streq(o[i], "vx"))  { 
            colvx   = need+1;   need += 1;      bits |= PhaseSpaceBit;
	} else if (streq(o[i], "vy"))  { 
            colvy   = need+1;   need += 1;      bits |= PhaseSpaceBit;
	} else if (streq(o[i], "vz"))  { 
            colvz   = need+1;   need += 1;      bits |= PhaseSpaceBit;
        } else if (streq(o[i], "phi"))  { 
            colphi  = need+1;   need += 1;      bits |= PotentialBit;
        } else if (streq(o[i], "acc"))  { 
            colacc  = need+1;   need += NDIM;   bits |= AccelerationBit;
	} else if (streq(o[i], "ax"))  { 
            colax   = need+1;   need += 1;      bits |= AccelerationBit;
	} else if (streq(o[i], "ay"))  { 
            colay   = need+1;   need += 1;      bits |= AccelerationBit;
	} else if (streq(o[i], "dens"))  { 
            coldens = need+1;   need += 1;      bits |= DensBit;
        } else if (streq(o[i], "az"))  { 
	    colaz   = need+1;   need += 1;      bits |= AccelerationBit;
        } else if (streq(o[i], "aux"))  { 
            colaux  = need+1;   need += 1;      bits |= AuxBit;
        } else if (streq(o[i], "key"))  { 
            colkey  = need+1;   need += 1;      bits |= KeyBit;
        } else warning("column name %s not understood",o[i]);
    }
    if (skip)
        dprintf(1,"%d: skipping block ; need=%d\n",id,need);
    else
        dprintf(1,
     "%d: mass(%d),pos(%d),vel(%d),phase(%d),phi(%d),acc(%d),aux(%d),dens(%d),key(%d)\n",
		id,colmass,colpos,colvel,colphase,colphi,colacc,colaux,coldens,colkey);

    
    for (j=0, nvals=0, bp=btab; j<nbody; j++, bp++) {
    	while (nvals < need) {

	  do {
            if (fgets(line, MAXLINE, instr) == NULL) {
	      if (j==0 && nvals==0) {
		warning("Block %d: line %d nvals=0",id,linecnt);
		return FALSE;
	      }
	      error("process(%s): unexpected EOF for particle %d",options,j+1);
	      /* in case error is bypassed: reset nbody */
	      nbody = j;
	      warning("Resetting nbody=%d",nbody);
	      return TRUE;
            } else {
	      if (Qcom && line[0]=='#') {
		dprintf(1,"COMMENT1: %s",line);
		if (Qhis) {
		  if (line[strlen(line)-1] == '\n') line[strlen(line)-1] = 0;        
		  shist[nhist++] = strdup(line);
		}
		continue;
	      } else
		break;
	    }
	  } while(line[0]=='#');    /* read until EOF or non-comment lines */

	  linecnt++;
	  if (line[strlen(line)-1] == '\n') line[strlen(line)-1] = 0;
#if 0            
	  if (skip) continue;
#else
	  if (skip) break;            
#endif            
	  if (Qcom) if (line[0] == '#') continue;
	  tab2space(line);
	  ngot = nemoinpd(line,&dvals[nvals],MAXVALS-nvals);
	  if (ngot < 0)
	    error("(%d) Line %d Parsing %s",ngot,linecnt,line);
	  else
	    nvals += ngot;
	  if (!wrap) break;
        }
        if (skip) continue;
        if (nvals > need) {
            if (wrap) {                 /* in wrap mode */
                nvals -= need;          
                if (spill)              /* only spill if asked for */
                    nvals = 0;
            } else                      /* else always spill */
                nvals = 0;
        } else if (nvals < need) {
	  dprintf(0," bad line:%s\n",line);
	  do {
	    fgets(line, MAXLINE, instr);
	    if (Qcom && line[0]=='#') {
	      dprintf(0,"COMMENT2: %s",line);
	      if (Qhis) {
		if (line[strlen(line)-1] == '\n') line[strlen(line)-1] = 0;        
		shist[nhist++] = strdup(line);
	      }
	    }
	  } while (line[0] == '#');
	      
	  dprintf(0,"next line:%s\n",line);
            error("Bad line %d (%d) Not enough values (need %d for %s) - try wrap",
                          linecnt,nvals,need,options);
        } else  /* exact match : nvals == need */
            nvals = 0;
                    
        if (colmass>0)   Mass(bp) = dvals[colmass-1];
        if (colpos>0)    SETV(Pos(bp),&dvals[colpos-1]);
        if (colx>0)      Pos(bp)[0] = dvals[colx-1];
        if (coly>0)      Pos(bp)[1] = dvals[coly-1];
        if (colz>0)      Pos(bp)[2] = dvals[colz-1];
        if (colvel>0)    SETV(Vel(bp),&dvals[colvel-1]);
        if (colvx>0)     Vel(bp)[0] = dvals[colvx-1];
        if (colvy>0)     Vel(bp)[1] = dvals[colvy-1];
        if (colvz>0)     Vel(bp)[2] = dvals[colvz-1];
        if (colphase>0) {
                         SETV(Pos(bp),&dvals[colphase-1]);
                         SETV(Vel(bp),&dvals[colphase-1+NDIM]);
                        }
        if (colphi>0)    Phi(bp) = dvals[colphi-1];
        if (colacc>0)    SETV(Acc(bp),&dvals[colacc-1]);
        if (colaux>0)    Aux(bp) = dvals[colaux-1];
        if (coldens>0)   Dens(bp)= dvals[coldens-1];
        if (colkey>0)    Key(bp) = dvals[colkey-1];
        if (j==0) dprintf(1,"%d: First line: %s\n",id,line);
        if (j==nbody-1) dprintf(1,"%d: Last line: %s\n",id,line);
    }
    dprintf(1,"%d: Linecount=%d\n",id,linecnt);
    return TRUE;
}


/*
 * GET_DOUBLE:  get 'n' double values from a header,
 *              no need to be on one line. 
 *
 * NOTE: not sure if the line on which last number (n) was
 *       read is still accessible, since I use fgets() in
 *       reading non-header information 
 *	 The extra \n seems to have done the trick.... not sure
 *	 about portability.... ought to use cashed fgets()+nemoinp()
 */

local int get_double(int n, double *d)
{
#if 1
						/* using fgets + nemoinp */
    int i, k=0;
    char line[MAXLINE];
    while (k<n) {
      do {
        if (fgets(line, MAXLINE, instr) == NULL) {
	  if (k==0) return 0;
	  warning("Unexpected EOF in header for snapshot %d",snapcnt+1);
	  return k;
        } else {
	  if (Qcom && line[0]=='#') {
	    dprintf(0,"COMMENT3: %s\n",line);
	    if (Qhis) {
	      if (line[strlen(line)-1] == '\n') line[strlen(line)-1] = 0;        
	      shist[nhist++] = strdup(line);
	    }
	    continue;
	  } else
	    break;
	}
      } while(line[0]=='#');    /* read until EOF or non-comment line */
      if (line[strlen(line)-1] == '\n') line[strlen(line)-1] = 0;        
      tab2space(line);
      i = nemoinpd(line,&d[k],n-k);
      if (i==0) return k;
      if (i<0) error("(%d) Error parsing %s",i,line);
      k += i;
    }
    return k;
#else
						/* using (slow) fscanf */
    int i;

    if (n <= 0) return n;
    for (i=0; i<n; i++) 
        if (fscanf(instr, " %lf\n", &d[i]) != 1) return 0;
    return n;
#endif
}

/*
 * GET_NBODY:   estimate nbody if none supplied  from the
 *		number of lines. Only works if no header
 *		given of course, since we don't know the
 *		number of lines a header consumes (see prev.
 *		routine get_double()
 *              Also doesn't work well when options=wrap would be needed
 */

local int get_nbody(void)
{
    if (hasvalue("header"))
        error("Need value for nbody=, or specify it in header=");
    nbody = nemo_file_lines(getparam("in"),0);
    if (nbody <= 0) 
        error("Cannot determine nbody, try nbody= or header=");
    if (nblocks == 0) 
        error("No block's specified");
    if (ntimes > 0) {
        if (nbody % ntimes != 0)
            warning("File-length (%d) not multiple of number of times (%d)",
                    nbody, ntimes);
        nbody /= ntimes;
    }
    if (nbody % nblocks != 0)
        warning("File-length (%d) not multiple of number of blocks (%d)",
            nbody, nblocks);
    nbody /= nblocks;
    return nbody;
}

/*
 * DO_SCAN: scan file, and report on regularity in number of columns
 */

void do_scan(stream instr)
{
    char line[MAXLINE];
    string *sp;
    int ncol, ncol_old=-1;

    warning("Scanning....");
    dprintf(0,"Line\t#Columns\n");
    while (fgets(line, MAXLINE, instr) != NULL) {
        linecnt++;
        if (line[strlen(line)-1] == '\n') line[strlen(line)-1] = 0;        
        sp = burststring(line," ,\t");
        ncol = xstrlen(sp,sizeof(string)) - 1;
        if (ncol_old != ncol) {
            printf("%d\t%d\n",linecnt,ncol);
            ncol_old = ncol;
        }
        freestrings(sp);
    }
    printf("%d\t%d\n",linecnt,ncol);
}


/*
 * small helper function, replaces tabs by spaces before processing.
 * this prevents me from diving into gipsy parsing routines and  fix
 * the problem there
 * PJT - June 1998.	 - see also tabmath.c
 */
        
local void tab2space(char *cp)
{
    while (*cp) {
        if (*cp == '\t') *cp = ' ';
        cp++;
    }
}

